### ISSUE TITLE
view::chunk does not work after view::join

### ISSUE THREAD

**Issue by FloopCZ** [ericniebler/range-v3#567]
Consider the following code:
```c++
#include <vector>
#include <range/v3/view.hpp>
using namespace ranges;

int main()
{
  std::vector<std::vector<int>> data{{1, 2, 3}, {4, 5, 6}};
  auto rng = data | view::join | view::chunk(3);
}
```
The compilation fails with the following static assertion: "The first argument to view::chunk must be a model of the ForwardRange concept". Does view::join somehow violate ForwardRange concept?

**Comment by FloopCZ** [comment]
I have examined the code and now I see the problem. view::join implements only the InputRange concept, because its cursors share a single "outer" and a single "inner" range stored inside the view object. If it should implement ForwardRange (multiple cursors allowed), each cursor would need its own "inner" range.

view::chunk requires ForwardRange argument, since in each increment, it performs two steps: take(n), advance(n). That's why the provided range has to support multiple cursors.

I see two solutions:
1. Store the "inner" range of the view::join inside the cursor (instead of inside the view object). For the cursor to be copyable and movable including its member iterators, the "inner" range would have to be a [view::shared](https://github.com/ericniebler/range-v3/pull/557) (so that the iterators would still be valid after copy/move).
2. Make a variant of view::chunk for InputRanges. That would require the chunk cursor to buffer the whole chunks, let's say in std::vector, and return the vector.

The first option is bad, since view::join would require dynamic heap allocation. The second option is even worse, since view::chunk iterator would contain the entire chunk. Can anyone see a better option?

**Comment by Victor-Savu** [comment]
@FloopCZ I wonder if solution 2 could be improved by providing an action (e.g. `action::buffer`) which would convert the InputRange into a ForwardRange by doing the buffering on-demand. I am not familiar with the internals (I would have to look at it this evening after work) but I can imagine that the buffering behavior can be extracted out of the implementation for your proposed second solution and put in a general buffering action. This way, heap allocation and buffering would happen, but it is at least explicit and keeps view::chunk lightweight.
I am not claiming this would be a good solution for this particular problem (not even that it makes option 2 any better than option 1), but if there is a feasible implementation it might solve a more general class of problems where implicit buffering is to be avoided.

Food for thought: is there a class of actions that can transform an InputRange into a ForwardRange? (i.e. a class which action::buffer would be a part of)

**Comment by FloopCZ** [comment]
@Victor-Savu I like your thinking! Unfortunately, to convert InputRange into ForwardRange in general will be slightly more difficult. InputRange is a SinglePass range, while ForwardRange can be iterated over and over. Therefore, if the user iterates through the whole hypothetical `action::buffer(inp_rng)`, the buffer will have to store all the elements of the `inp_rng`, just for the case that the user will want to iterate it again. Effectively, the easiest way to convert InputRange to ForwardRange is to convert it into a container (or is there anything better?).

The `view::chunk` for InputRanges would do something slightly better, since it would only "containerize" one chunk at a time.

**Comment by gnzlbg** [comment]
> Effectively, the easiest way to convert InputRange to ForwardRange is to convert it into a container (or is there anything better?).

Is there any other way? To convert an InputRange that is not ForwardRange into a ForwardRange you need to store the elements of the range somewhere. 

I actually think that the user should be in charge of deciding where to store these elements and that should be explicit. `action::push_back/insert` already allow this, so I am wondering why do we need `action::buffer` ?

**Comment by FloopCZ** [comment]
@gnzlbg Exactly, `action::buffer` would have no purpose. We were just thinking out loud what are the alternatives for allowing view::chunk and view::join to work together.

**Comment by Victor-Savu** [comment]
@gnzlbg Yes, you are very right! To my shame, I am not that familiar with the library. I should have thought that some form of buffering was already available.

**Comment by FloopCZ** [comment]
One more idea. `view::chunk` (and also `view::group_by` etc.) may have a variant for InputRanges implemented as following. In the read() function, the `view::chunk::cursor` returns a special range called e.g., `chunk_range`. The `view::chunk::cursor` and the `chunk_range::cursor`, share a single iterator to the underlying range. This shared iterator is stored in the `view::chunk` object. If the `chunk_range::cursor` is incremented, it simply increments the shared iterator by one. If the `view::chunk::cursor` is incremented, it keeps incrementing the shared iterator until the end of the current chunk (because user could only iterate the first few elements of the chunk and leave the rest intact).

This one is my favourite so far. No heap allocation, no storage. One just has to be careful - when the `view::chunk::cursor` is iterated, the `chunk_range::cursor` invalidates. On the other hand, that is a basic property of InputRanges. Furthermore, it is possible to produce a runtime assertion if the user violates this rule (the chunks can be counted in the `view::chunk` and the `chunk_range` can know the chunk number it belongs to). What do you think?

**Comment by gnzlbg** [comment]
This might be related to #191, which contains an explanation about _why_ `view::join` returns an `InputRange` (and why it must be `SinglePass`).

**Comment by FloopCZ** [comment]
This thread starts to be confusing. Just to sum up, there are two separate things to be discussed.

1. Implementation of view::chunk and view::group_by for InputRanges as [a mutable-only InputRange](https://github.com/ericniebler/range-v3/issues/567#issuecomment-283206142).

2. How to distinct true single pass InputRanges (which cannot be copied), from lazy InputRanges, which can be copied (such as view::join). It is more thoroughly explained in the comment above. I suggest @gnzlbg to create a separate issue for this.

**Comment by gnzlbg** [comment]
Yes, sorry about that. Maybe we can rename this issue to "view::chunk / view::group_by should work on InputRanges"? I mean, the root of the problem right now is that view::join returns an InputRange and view::chunk/group_by don't work on those.

**Comment by pzelasko** [comment]
I think I have a related problem when trying to use `view::chunk` with `istream_range`. Consider following snippet:

    void foo() {
      std::ifstream ifs{"myfile", std::ios::binary};
      auto bytes = ranges::istream<unsigned char>(ifs);
      auto byte_pairs = bytes | view::chunk(2);
    }

which does not compile in clang 3.8 with `-std=c++14` [giving this error message](https://pastebin.com/u7Lrcdrz). Or did I miss something?

**Comment by ericniebler** [comment]
@pzelasko as of 6f1e194, `view::chunk` works with Input views.

**Comment by pzelasko** [comment]
You're right, thanks. FYI GCC 7.1 error message was a bit more clear - I needed to `move` the `bytes` Input range (in case anyone stumbles upon this).

**Comment by CaseyCarter** [comment]
>  I needed to move the bytes Input range (in case anyone stumbles upon this)

This is a bug: the input-range specialization of `chunk_view` was not properly accepting lvalue ranges. (I know; appalling lack of test coverage.)
